compiler/rustc_resolve/src/macros.rs

```diff
@@ -708,15 +736,29 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                 // expanded into a dummy fragment for recovery during expansion.
                 // Now, post-expansion, the resolution may succeed, but we can't change the
                 // past and need to report an error.
-                // However, non-speculative `resolve_path` can successfully return private items
+                // Special cases:
+                // 1. non-speculative `resolve_path` can successfully return private items
                 // even if speculative `resolve_path` returned nothing previously, so we skip this
                 // less informative error if the privacy error is reported elsewhere.
+                // 2. issue #118455, unresolved top level attribute error didn't imported prelude and
+                // already have emitted an error, report builtin macro and attributes error by the way,
+                // so `check_invalid_crate_level_attr` in can ignore them.
+
                 if this.privacy_errors.is_empty() {
-                    this.dcx().emit_err(CannotDetermineMacroResolution {
-                        span,
-                        kind: kind.descr(),
-                        path: Segment::names_to_string(path),
-                    });
+                    if this.is_builtin_macro(res)
+                        || this.builtin_attrs_bindings.values().any(|b| b.res() == res)
+                    {
+                        if this.report_invalid_crate_level_attr(&krate.attrs, path[0].ident.name) {
+                            return;
+                        }
+                        if this.tcx.dcx().has_errors().is_none() {
```

Comment:
This change is orthogonal to the rest of the PR, it has nothing to do with crate-level attributes.

I think we can skip following `CannotDetermineMacroResolution` errors if one `CannotDetermineMacroResolution` was reported.
But it's better done in a separate PR.

Reply:
On a second look, it's fine to skip `CannotDetermineMacroResolution` if *any* other error was reported.
They are useless by themselves, and only need to be reported if the compilation otherwise succeeds.

Reply:
It is even possible that this error can be turned into `span_delayed_bug`, could you check?

Reply:
yes, this is a simple fix for #118455, I can land it separately.

my crate-level logic here is also trying to resolve this issue (#89566, #67107)

```rust
#![derive(Debug)]
struct Test {
    s: String,
}

fn main() {
}

```

the output is:
```
error: cannot determine resolution for the attribute macro `derive`
 --> ./t/derive.rs:1:4
  |
1 | #![derive(Debug)]
  |    ^^^^^^
  |
  = note: import resolution is stuck, try simplifying macro imports

error: `derive` attribute cannot be used at crate level
 --> ./t/derive.rs:1:1
  |
1 | #![derive(Debug)]
  | ^^^^^^^^^^^^^^^^^
2 | struct Test {
  |        ---- the inner attribute doesn't annotate this struct
  |
help: perhaps you meant to use an outer attribute
  |
1 - #![derive(Debug)]
1 + #[derive(Debug)]
  |
```

the crate-level logic is trying to report only:

```console
error: `derive` attribute cannot be used at crate level
 --> ./t/derive.rs:1:1
  |
1 | #![derive(Debug)]
  | ^^^^^^^^^^^^^^^^^
  |
help: perhaps you meant to use an outer attribute
  |
1 - #![derive(Debug)]
1 + #[derive(Debug)]
  |
```

Reply:
> It is even possible that this error can be turned into `span_delayed_bug`, could you check?

ah, seems a good idea, maybe we can solve these two kinds of issue!
let me check.

Reply:
only this case can not be handled:

```rust
use std as derive;

#[derive(Default)] //~ ERROR cannot determine resolution for the attribute macro `derive`
struct S;

fn main() {}

```

Reply:
let's stash this error!
it will reported if no other errors reported, seems work like a charm.